### ISSUE TITLE
localmarks option weirdly redefines \markright

### ISSUE THREAD

**Issue by jfbu** [reutenauer/polyglossia#396]
This is a spin-off from #395.

Currently after main #395 has been fixed, and better indentation added, https://github.com/reutenauer/polyglossia/blob/3d81642bb5ab8942061aad848439e3f071135598/tex/polyglossia.sty#L979-L995

But if we follow what happens we see that it add the polyglossia patch (as a wrapper) thrice: twice for the right mark and once for the left mark. I believe it should be only once for the right mark. And arguably none for the left mark, because the LaTeX2e definition in source2e of `\markright` leaves the left mark intact.

Thus modifying the left is not quite the same as modifying the behaviour of `\markright` it also uses it in a kind of a posteriori fix things manner which does not sound right.

(by left and right mark I mean the respectively first and second braced groups of LaTeX2e internal `\@themark`)

**Comment by jfbu** [comment]
The fix consists in simply avoiding redefining the `\@markright` LaTeX internal.
```diff
diff --git a/tex/polyglossia.sty b/tex/polyglossia.sty
index 99160fc..af0798b 100644
--- a/tex/polyglossia.sty
+++ b/tex/polyglossia.sty
@@ -986,13 +986,6 @@
             \endgroup
             \if@nobreak\ifvmode\nobreak\fi\fi%
          }%
-         \def\@markright####1####2####3{%
-            \@temptokena{\protect\@@ensure@maindir{####1}}%
-            \unrestored@protected@xdef\@themark{%
-               {\the\@temptokena}%
-               {\protect\@@ensure@maindir{####3}}%
-            }%
-         }%
       }%
 }
 
```

Indeed  `\@markright` is only used once in source2e

```
$ grep -C2 @markright `kpsewhich latex.ltx`
  \begingroup
    \let\label\relax \let\index\relax \let\glossary\relax
    \expandafter\@markright\@themark {#1}%
    \@temptokena \expandafter{\@themark}%
    \mark{\the\@temptokena}%
  \endgroup
  \if@nobreak\ifvmode\nobreak\fi\fi}
\def\@markright#1#2#3{\@temptokena {#1}%
  \unrestored@protected@xdef\@themark{{\the\@temptokena}{#3}}}
\def\leftmark{\expandafter\@leftmark\botmark\@empty\@empty}
```
Its role is to handle the `#3` which will be "right mark" via a protected xdef, but it leaves intact the `#1`.
The `#2` is former "right mark" and is replaced by `#3`.

In the polyglossia code the re-def'ed `\@markright` does overzealous things. The `#3` has already been wrapped.

Thus, simply using the LaTeX2e official `\@markright` looks right.

**Comment by jfbu** [comment]
> Thus, simply using the LaTeX2e official `\@markright` looks right.

However in case for some reason one wants `\markright` to also modify the stored left part of the `\@themark` LaTeX2e internal macro, here is how to do it:

```diff
diff --git a/tex/polyglossia.sty b/tex/polyglossia.sty
index 99160fc..881dd90 100644
--- a/tex/polyglossia.sty
+++ b/tex/polyglossia.sty
@@ -987,10 +987,10 @@
             \if@nobreak\ifvmode\nobreak\fi\fi%
          }%
          \def\@markright####1####2####3{%
-            \@temptokena{\protect\@@ensure@maindir{####1}}%
+            \@temptokena{\lowercase{\foreignlanguage{##1}}{\protect\@@ensure@maindir{####1}}}%
             \unrestored@protected@xdef\@themark{%
                {\the\@temptokena}%
-               {\protect\@@ensure@maindir{####3}}%
+               {####3}%
             }%
          }%
       }%
```

Successive usage of `\markright` will nest more and more the "left mark". This does not sound quite right.

**Comment by jspitz** [comment]
So you suggest your first patch, right? I know next to nothing about the idea behind this whole option.

**Comment by jfbu** [comment]
@jspitz yes I think the normal thing to do is the first patch, i.e. simply remove the redefinition of `\@markright`. This is advice provided without any warranty nor liability of any kind for any purpose. Seriously, this is untested and only from a "Vulcan" point of view of logic.

**Comment by jspitz** [comment]
This is > 10 year old code (written by the first polyglossia maintainer, François Charette), so I suppose our best bet is to remove/comment out that code and see if somebody screams.

@reutenauer what do you think?

**Comment by reutenauer** [comment]
Agreed.

**Comment by jspitz** [comment]
OK, done at b446b4d1b84b8.

**Comment by jspitz** [comment]
Another issue that struck me while looking at the code: LaTeX nowadays defines `\markboth` and `\markright` as robust commands via `\DeclareRobustCommand`. I wonder whether we should do this as well in `\enable@local@marks`

**Comment by jspitz** [comment]
> Another issue that struck me while looking at the code: LaTeX nowadays defines `\markboth` and `\markright` as robust commands via `\DeclareRobustCommand`. I wonder whether we should do this as well in `\enable@local@marks`

I did that now at 78f58f830955ee and also used l3 syntax. @jfbu I'd appreciate if you could double check whether this breaks something for you.

**Comment by jfbu** [comment]
looks good to me

```latex
\documentclass{book}
\usepackage[localmarks]{polyglossia}
\setmainlanguage{english}
\setotherlanguage{french}

\usepackage{fancyhdr}
\fancyhead{}
\fancyhead[R]{\MakeUppercase{L=(\leftmark), R=(\rightmark)}}

\makeatletter
\def\showmark{\show\@themark}
\makeatother
\begin{document}
\tableofcontents
%\showmark

\chapter{First}
\pagestyle{fancy}

A
\section{A}

A

\section{B}

A\showhyphens{constitution}
\markright{last left mark in chapter 1!}
%\showmark

\selectlanguage{french}
\chapter{Deuxième}

A

\section{A}

A

\section{B}

A
\clearpage

B\showhyphens{constitution}
\markright{dernier mark droit dans le chapitre 2? \lowercase{\showhyphens{constitution}}}
%\showmark
\end{document}
```
